pkg/directory/gh.go

```diff
@@ -0,0 +1,363 @@
+package directory
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/google/go-github/github"
+	"github.com/pkg/errors"
+	"golang.org/x/oauth2"
+)
+
+const (
+	// GHAllTeam containing all users in GitHub
+	GHAllTeam = "all"
+)
+
+// UsersService holds methods used in the GitHub UsersService for easier testing
+type UsersService interface {
+	Get(context.Context, string) (*github.User, *github.Response, error)
+}
+
+// GH hosts a client for accessing GH as well as cached Member and Team lists
+type GH struct {
+	*github.Client
+
+	UsersService UsersService
+	Info
+}
+
+// NewGitHub returns an initialized GitHub client to the caller and stored GH members and teams
+func NewGitHub(org string) (*GH, error) {
+	ctx := context.Background()
+	client := &GH{}
+
+	token, ok := os.LookupEnv("GITHUB_TOKEN")
+	if !ok {
+		return client, errors.New("GITHUB_TOKEN not set")
+	}
+	ts := oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: token},
+	)
+	tc := oauth2.NewClient(ctx, ts)
+	client.Client = github.NewClient(tc)
+	client.UsersService = client.Client.Users
+	client.Org = org
+
+	if err := client.getMembersAndTeams(); err != nil {
+		return client, err
+	}
+	return client, nil
+}
+
+func (g *GH) getMembersAndTeams() error {
+	update := false
+
+	_, err := os.Stat(cacheDir)
+	if err != nil {
+		os.MkdirAll(cacheDir, os.ModePerm)
+	}
+
+	membersFile := path.Join(cacheDir, "members")
+	mfInfo, err := os.Stat(membersFile)
+	if err != nil || time.Since(mfInfo.ModTime()).Minutes() > cacheTTL {
+		update = true
+	}
+
+	teamsFile := path.Join(cacheDir, "teams")
+	tfInfo, err := os.Stat(teamsFile)
+	if err != nil || time.Since(tfInfo.ModTime()).Minutes() > cacheTTL {
+		update = true
+	}
+
+	if update {
+		wg := sync.WaitGroup{}
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			g.getMembers()
+		}()
+
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			g.getTeams()
+		}()
+		wg.Wait()
+
+		if err := saveCache(membersFile, g.Members); err != nil {
+			return err
+		}
+		if err := saveCache(teamsFile, g.Teams); err != nil {
+			return nil
+		}
+	} else {
+		if err := getCached(membersFile, &g.Members); err != nil {
+			return err
+		}
+		if err := getCached(teamsFile, &g.Teams); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func saveCache(filename string, v interface{}) error {
+	buf, err := json.Marshal(v)
```

Comment:
Consider using sqlite (or a pure Go alternative) vs flat JSON files.

Reply:
I don't really understand this. SQLite has a CGO requirement which would make cross compilation annoying. Otherwise, this file isn't big and I don't see what this would really gain us since this will only be access a few times between cache invalidation when a user is trying to share a secret.

Reply:
Whenever somebody writes data to a local file I think sqlite (or something similar, like you said, due to CGo) should be considered. Especially if the file will be updated or modified in the future. If that isn't the case here, feel free to disregard this comment.

Reply:
This file is overwritten after the TTL (1 hour) is reached with new information. So there isn't any updating and if the structures are changed then the entire file will be rewritten.

Reply:
OK. 👍